lib/AnalysisStructured/PtrAnalysis.cpp

```diff
@@ -0,0 +1,1128 @@
+    newOp.getOperation()->print(llvm::dbgs(),
+                                OpPrintingFlags().printGenericOpForm());
+    llvm::dbgs() << "\n";
+  });
+
+  op->erase();
+  return success();
+}
+
+LogicalResult PtrAnalysis::rewriteLoadOp(triton::LoadOp op) {
+  auto ptr = map.lookupOrNull(op.getPtr());
+  auto mask = op.getMask();
+  auto other = op.getOther();
+  auto loc = op.getLoc();
+
+  if (!ptr) {
+    op->emitRemark("PtrAnalysis: pointer is not replace with tts.make_tptr so "
+                   "loadOp cannot be rewritten");
+    return failure();
+  }
+
+  if (!ptr.getType().isa<RankedTensorType>()) {
+    op->emitRemark("PtrAnalysis: scalar loadOp will not be rewritten");
+    return failure();
+  }
+
+  ArrayRef<OpFoldResult> dims;
+  MaskSState mstate;
+  Value scalarOther;
+
+  OpBuilder builder(op);
+  // Analyze the mask operand to determine at runtime the size of the data we
+  // are moving.
+  if (mask) {
+    if (mstate.parse(mask, loc, builder).failed()) {
+      op->emitRemark("MaskAnalysis failed");
+      return failure();
+    }
+    dims = mstate.dims;
+  }
+
+  if (other) {
+    assert(mask && "other value used while no masks are specified");
+
+    scalarOther = getScalarValue(other, loc, builder);
+    if (!scalarOther) {
+      op->emitRemark("other value used in masked load produced by "
+                     "unsupported instruction");
+      return failure();
+    }
+  }
+
+  auto loadOp = builder.create<tts::LoadOp>(loc, ptr, dims, scalarOther);
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "creating tts::load:\n";
+    loadOp->dump();
+  });
+
+  op.replaceAllUsesWith(loadOp.getResult());
+  op->erase();
+  return success();
+}
+
+LogicalResult PtrAnalysis::rewriteStoreOp(triton::StoreOp op) {
+  auto ptr = map.lookupOrNull(op.getPtr());
+  auto val = op.getValue();
+  auto mask = op.getMask();
+  auto loc = op.getLoc();
+
+  if (!ptr) {
+    op->emitRemark("PtrAnalysis: pointer is not replace with tts.make_tptr so "
+                   "storeOp cannot be rewritten");
+    return failure();
+  }
+
+  if (!ptr.getType().isa<RankedTensorType>()) {
+    op->emitRemark("PtrAnalysis: scalar loadOp will not be rewritten");
+    return failure();
+  }
+
+  ArrayRef<OpFoldResult> dims;
+  MaskSState mstate;
+
+  OpBuilder builder(op);
+
+  // Analyze the mask operand to determine at runtime the size of the data
+  // are moving.
+  if (mask) {
+    if (mstate.parse(mask, loc, builder).failed()) {
+      op->emitRemark("MaskAnalysis failed");
+      return failure();
+    }
+    dims = mstate.dims;
+  }
+
+  auto storeOp = builder.create<tts::StoreOp>(loc, ptr, val, dims);
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "creating tts::store:\n";
+    storeOp->dump();
+  });
+
+  op->erase();
+  return success();
+}
+
+LogicalResult PtrAnalysis::rewriteOp(Operation *rootOp) {
+  LLVM_DEBUG({
+    llvm::dbgs() << "rewriting rootOp\n";
+    rootOp->dump();
+  });
+
+  rootOp->walk<WalkOrder::PreOrder>([&](Operation *op) {
```

Comment:
> triton-to-structured does not use DialectConversion but rather manually walks the IR to perform rewriting. The reason is this pass does not try to legalize Triton operations (in fact it allows them to live after the pass) but rather opportunistically rewrite certain ops if the analysis succeeds. Legality marking in DialectConversion does not provide a way to set certain instances of an op as legal, which makes it a poor fit for our purpose.

I believe in this case we can make use of the [Pattern Rewrite Driver](https://mlir.llvm.org/docs/PatternRewriter/#greedy-pattern-rewrite-driver) (`applyPatternsAndFoldGreedily`). The driver will repeatedly call these patterns until the patterns fail to match (i.e.: `return failure()`). I think this fits quite well with most of the rewrite functions that we have here. `rewriteForOp` might be a trickier because now we need to determine if the `scf::ForOp` has already been rewritten. Would be nice if we can use this, but otherwise if it's too much complexity, I think it's fine to use the approach we have here.

Reply:
That's a very good point. `rewriteForOp` was the thing that gave me most headache. The reason is that it can potentially produce operations that we may or may not want to rewrite. With `DialectConversion` it's not easy to say "I tried rewriting this Op but failed and I am ok with leaving it as is". It looks like `PatternRewriter` is probably a better fit. But tbh it takes so much time to try something like this I kinda don't want to spend the time.

Another potentially tricky thing is we don't always want to replace the results of `addptr`. E.g., if we have a chain of addptr->addptr->load. We can rewrite the first addptr but not the second one, we'd like to keep the first one as is. It's unclear if `PatternRewriter` can handle all this. I believe the manual walk + custom data structure provides the most flexibility. (This is currently handled by the additional IRMapping defined in PtrAnalysis struct.)

Reply:
This sounds good to me. Not worth it to spend time catering to the PatternRewriter.